### ISSUE TITLE
binary compatibility between librsync 1.0.0 and librsync 2.0.0

### ISSUE THREAD

**Issue by stephane78** [librsync/librsync#97]
Hi,

I have a question is librsync 2.0.0 binary compatible with 1.0.0, i.e are  the delta and signatures file generated by librsync 1.0.0 compatible with librsync 2.0.0 ? because we use now librsync 1.0.0 but perhaps should upgrade the version....

**Comment by stephane78** [comment]
Hi, the fopen of the signature to create the signature file returns 0 on this files so I suppose it is the problem (encoding problem) but if we convert to utf8, fopen don't open the file.

**Comment by stephane78** [comment]
It is not a pb of librsync. more a problem of fopen of mingw-w64 .but for my second question are the deltas and the signatures files of librsync 2.0.0 the same as those of librsync 1.0.0 or are the signatures and deltas files of librsync 1.0.0 compatible with librsync 2.0.0

**Comment by stephane78** [comment]
(my first problem is solved thanks to the function _wfopen)

**Comment by dbaarda** [comment]
I've not tested to confirm it works as intended, but the signature and delta files should be forwards compatible. So signature and delta files generated by 1.0.0 should work fine with 2.0.0.

It should also be possible to generate signature and delta files with 2.0.0 that work with 1.0.0 by providing the right arguments. In particular you probably need to select md4sum signatures, not blake2.

**Comment by stephane78** [comment]
ok.But we already use Blake2 signatures with 1.0.0.I think we must test it to see if it is compatible....

**Comment by dbaarda** [comment]
I've done some compatibility tests between the rdiff as distributed with Debian;

$ rdiff -V
rdiff (librsync 0.9.7) [x86_64-pc-linux-gnu]

and a new rdiff built from head as a RELEASE build with trace turned on;

$ ./rdiff -V
rdiff (librsync 2.0.1)

I ran the attached [version.test](https://github.com/librsync/librsync/files/1043423/version.test.txt) script from the librsync/tests directory;

11:52:36 abo@minkirri:~/src/librsync/librsync/tests
$ ./version.test .. 64K ~/tmp/new >~/tmp/result.txt 2>&1

which gave the attached [result.txt](https://github.com/librsync/librsync/files/1043416/result.txt). Things to note;

* The new rdiff requires -f to overwrite existing files.
* The only combination that didn't work (as expected) was using the old rdiff with a new blake2 signature.
* The old sig is identical to a new md4 sig (Note I tried this different and missing -S arguments).
* The delta files produced by old and new versions using any (working combination) sig are identical.
* The new rdiff is 2.5x as fast for delta operations using either kind of strongsum.

So I think this pretty conclusively proves that they are compatible in every way except using a blake2 sig with a pre 1.0 version.

I haven't got/tried a 1.0.0 version where blake2 was introduced, but I'd be extremely surprised if it wasn't 100% compatible with current head.
